### PR TITLE
Optimize SM4 by AES-NI/AVX in x86_64

### PR DESCRIPTION
**Pull request by HowjuLee**
# Optimize SM4 by AES-NI/AVX in x86_64
    
    This patch adds AES-NI/AVX/x86_64 assembler implementation of SM4
    block cipher. Through two affine transforms, we can use the AES S-Box
    to simulate the SM4 S-Box to achieve the effect of instruction
    acceleration. Because the parameter passing is different between
    Unix and Win64, Win64 version is remained to be later completed.
    
    The main algorithm implementation comes from SM4 AES-NI work by
    libgcrypt and Markku-Juhani O. Saarinen at:
    https://github.com/mjosaarinen/sm4ni
    
    This optimization supports the four modes of SM4, ECB, CBC, and
    CTR. Since CBC do not support multiple block parallel
    encryption, the optimization effect is not obvious.
    
    Benchmark on Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz. The abscissas are input of different bytes size.
    The 'numbers' are in 1000s of bytes per second processed.
    
    type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
    generic
    SM4-ECB         117356.68k   127807.49k   130309.38k   132134.01k   133133.65k   132514.45k
    SM4-CBC         111685.46k   120877.27k   119651.75k   117984.04k   118308.42k   120952.15k
    SM4-CTR         116462.33k   119879.49k   121615.87k   121914.71k   121132.37k   120321.03k
    
    Sm4-aesni-avx(2.x times faster)
    SM4-ECB          60237.89k   270777.97k   294322.35k   304375.13k   299098.11k   289875.84k
    SM4-CBC          59606.40k    57613.14k    58730.06k    58581.05k    59683.00k    56645.76k
    SM4-CTR          55471.44k   264135.88k   275725.74k   282237.95k   285656.96k   283142.83k
    
Signed-off-by: Haoru Li <lihaoru@linux.alibaba.com>

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
Looks like newly added workflows run (or try to run) even when Github ones are not confirmed...

@openssl/committers

**Comment by HowjuLee** [comment]
> Looks like newly added workflows run (or try to run) even when Github ones are not confirmed...
> 
> @openssl/committers

I just realize that I can use the Action in my own repo to check all workflows, which makes my early attempt to cross-compile on all platform in my own environment a little dump.

**Comment by zzl360** [comment]
the benchmark data is different from https://github.com/torvalds/linux/commit/5b2efa2bb865eb784e06987c7ce98c3c835b495b

File: crypto/sm4/asm/aesni_sm4-x86.pl [hunk 1]
```diff
@@ -0,0 +1,1119 @@
...
+
+@_4args=$win64?	("%rcx","%rdx","%r8", "%r9") :	# Win64 order
+		("%rdi","%rsi","%rdx","%rcx");	# Unix order
+
+# $PREFIX = "_sm4_aesni";
+$PREFIX = "aesni_sm4";
+
+{{{
+$code.=<<___;
+.arch x86-64
```
Comment (zzl360) [inline crypto/sm4/asm/aesni_sm4-x86.pl:44]: x86_64?

Comment (zzl360) [reply crypto/sm4/asm/aesni_sm4-x86.pl:44]: should delete this line

**Pull request closed without merging**